Simplify and improve Dynamic Notch

ctzsnooze opened the pull request:
This is a set of changes to the previous Dynamic Notch code.

First a huge thank you to @rav-rav for the original code and @kmitchel for his advice, assistance and enthusiasm.

I am putting them up here for testing and comment.  I have annotated extensively and summarised how the code works.  Performance is unlikely to be worse than the old code, but may not be much better, depending on the quad.

If you are a coder please look at the queries I put at the top of the file.

When testing, please use `set debug_mode = FFT_FREQ`, make a log using a current 4.2 beta, and a second log with no other changes other than this revised dynamic notch code.  If possible, do two flights on each code base, one with rpm on, one with rpm off (four flights in all).  

The code still seeks whatever peak is highest, but has a slightly different, more accurate, and hopefully less CPU intensive method of converting that peak to frequency.  

It takes about 24ms to fully exchange the historical data values in the circular FFT buffer, so the FFT does not track 'instantly'.  Additionally, if FFT / Dynamic Notch suppresses a resonant peak, it typically will move somewhere else, only to have that peak return.  This caused the older code to 'yo-yo' back and forth between peaks.   

This code uses a dynamic lowpass filter where usually the movement from one frequency centre to the next is slow - a 5hz first order lowpass.  However the cutoff frequency can be dynamically made faster, up to 5x faster, when the peak is big.  The larger the peak, the quicker the centre frequency of the FFT will move towards the peak.  Conversely, if the peak is suppressed, and an alternate peak exists that is small, the FFT will only slowly move away from the large peak.  This tends to stabilise the FFT in the region of larger peaks much better than before.  It can be tweaked somewhat.  The min cutoff is `#define DYN_NOTCH_SMOOTH_HZ 5` ie 5hz.  The max increase in filter movement above this is set by `float dynamicFactor = constrainf(dataMax / dataMin, 1.0f, 5.0f);`

This is a brief summary of changes:

Remove prevCenterFreq[axis] (didn't achieve much since centerFreq is constantly changing)
Simpler centre frequency detection:
- cubes, not squares
- only look one bin either side
Bin start and max values checked and kept within limits (not use bin 0 or >15)
Updated annotations and explanations
Whitespace fixes
Dynamic PT1 smoothing of centerFreq, more quickly to larger peaks, slower to smaller peaks
biquad filter on centerFreq removed (replaced with PT1)
centerFreq's are now floats

Testing and feedback gratefully appreciated.

ctzsnooze commented:
Added some small updates to the peak detection logic to ensure it would never evaluate bin 0 (the 'dc' bin) and to get the nearest local minimum before the biggest peak in the FFT, rather than the first minimum it came across.  This is likely to better determine the 'height' of a peak.

ctzsnooze commented:
Some examples, these with rpm off and PIDs 0, to generate a significant motor rpm related incoming sine wave.  4k4k so updates to centerFreq every 3ms.

Upper trace is motors, middle yellow line is dynamic notch centre frequency, middle blue line is peak height (constrained 100-500, where 500 means the peak was 5x higher than the previous minimum).

Lower two traces are gyro (yellow, post dynamic notch) and gyro (red, pre-dynamic notch).

First spectrum is from pre-dynamic notch showing a ton of rpm related noise.

![Screen Shot 2020-01-02 at 11 52 41](https://user-images.githubusercontent.com/11737748/71649268-428f8100-2d61-11ea-971d-e1c1b24c9fc5.jpg)

Second image is the same data but the spectrum is after the dynamic notch...
![Screen Shot 2020-01-02 at 11 51 06](https://user-images.githubusercontent.com/11737748/71649281-64890380-2d61-11ea-8902-8e44a0e6b1d3.jpg)

With rpm on and PIDs on, this quad only had weak resonance lines.  The residual peaks were not strong, so the peak height was a bit random.  Looked at very closely, it came up whenever there was a decent bit of resonance to deal with.  

First log has a spectrum from before the dynamic notch, showing a bump around 180-250hz from various resonances or non-rpm noise.
![Screen Shot 2020-01-02 at 11 57 16](https://user-images.githubusercontent.com/11737748/71649312-afa31680-2d61-11ea-9dce-62a474a18451.jpg)

Next image shows post-dynamic notch on same file, those peaks are largely gone, leaving a more monotonic decline in noise...
![Screen Shot 2020-01-02 at 11 57 28](https://user-images.githubusercontent.com/11737748/71649319-cb0e2180-2d61-11ea-9dea-0ea6dea7b3da.jpg)

This image is an annotated spectrum, showing the small but definite reduction in resonance lines and 3rd harmonic removal.  Pre-dynamic on left, post-dynamic on right.
![Screen Shot 2020-01-02 at 13 15 45](https://user-images.githubusercontent.com/11737748/71649348-11638080-2d62-11ea-81ec-cf8107f4171c.jpg)

Asizon commented:
I have finished my tests. 3 flights with new test code and 3 flights with current 4.2 Master. There is a diff with your recommended test settings:
```
# master
set dyn_notch_width_percent = 0
set dyn_notch_q = 200
set dyn_notch_min_hz = 60
set acc_hardware = NONE
set mag_hardware = NONE
set baro_hardware = NONE
set serialrx_provider = FPORT
set serialrx_inverted = ON
set serialrx_halfduplex = ON
set dshot_bidir = ON
set motor_pwm_protocol = DSHOT300
set vbat_warning_cell_voltage = 320
set pid_process_denom = 2
set osd_vbat_pos = 2497
set osd_rssi_pos = 2074
set debug_mode = FFT_FREQ
set gyro_1_align_yaw = 1800
set gyro_2_bustype = NONE
set name = ASZN
```
There are 3 TestCode logs:
[TESTCODE4.2.zip](https://github.com/betaflight/betaflight/files/4015546/TESTCODE4.2.zip)
And there 3 Master4.2 logs:
[MASTER4.2.zip](https://github.com/betaflight/betaflight/files/4015548/MASTER4.2.zip)
I hope this will help @ctzsnooze ;)

Asizon commented:
I have finished my tests. 3 flights with new test code and 3 flights with current 4.2 Master. There is a diff with your recommended test settings:
```
# master
set dyn_notch_width_percent = 0
set dyn_notch_q = 200
set dyn_notch_min_hz = 60
set acc_hardware = NONE
set mag_hardware = NONE
set baro_hardware = NONE
set serialrx_provider = FPORT
set serialrx_inverted = ON
set serialrx_halfduplex = ON
set dshot_bidir = ON
set motor_pwm_protocol = DSHOT300
set vbat_warning_cell_voltage = 320
set pid_process_denom = 2
set osd_vbat_pos = 2497
set osd_rssi_pos = 2074
set debug_mode = FFT_FREQ
set gyro_1_align_yaw = 1800
set gyro_2_bustype = NONE
set name = ASZN
```
There are 3 TestCode logs:
[TESTCODE4.2.zip](https://github.com/betaflight/betaflight/files/4015601/TESTCODE4.2.zip)
And there 3 Master4.2 logs:
[MASTER4.2.zip](https://github.com/betaflight/betaflight/files/4015603/MASTER4.2.zip)
I hope this will help @ctzsnooze ;)

rav-rav commented on src/main/flight/gyroanalyse.c:
```diff
@@ -112,13 +152,15 @@ void gyroDataAnalyseInit(uint32_t targetLooptimeUs)
+    fftResolution = (float)fftSamplingRateHz / FFT_WINDOW_SIZE; // 41.65hz per bin for medium
 
-    fftResolution = (float)fftSamplingRateHz / FFT_WINDOW_SIZE;
+    fftStartBin = MAX(2, dynNotchMinHz / lrintf(fftResolution)); // can't use bin 0
```
Why do you also leave out bin 1?

ctzsnooze replied:
Bin 1 is almost always still carrying control information.  I do compare bin 2 to bin 1 if bin 2 is max.  I would not want to compare bin 1 to bin 0, if bin 1 was max.  Setting the start bin to 2 was the easiest way I found to achieve this.  

Bin 0 is DC value of the entire FFT block, not a frequency bin.  When I logged it, it held erratic values that were useless for comparisons of AC amplitudes.

Balint134 replied:
Can't this `MAX()` just be removed to avoid confusion? It doesn't really matter at all because in [settings.c](https://github.com/betaflight/betaflight/blob/master/src/main/cli/settings.c#L659) there's a minimum 60Hz value anyways. With this resolution we have here 60Hz will fall to bin 1.

rav-rav commented on src/main/flight/gyroanalyse.c:
```diff
@@ -138,14 +180,12 @@ void gyroDataAnalyseStateInit(gyroAnalyseState_t *state, uint32_t targetLooptime
     arm_rfft_fast_init_f32(&state->fftInstance, FFT_WINDOW_SIZE);
 
 //    recalculation of filters takes 4 calls per axis => each filter gets updated every DYN_NOTCH_CALC_TICKS calls
+//    at 4khz gyro loop rate this means 8khz / 4 / 3 = 666Hz => update every 1.5ms
```
"at 8kHz gyro loop rate"

Actually all khz should be replaced by kHz.

ctzsnooze replied:
Thanks I'll do that - my bad

rav-rav commented on src/main/flight/gyroanalyse.c:
```diff
@@ -269,62 +311,64 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state,
+            if (binMax == 0) { // no bin increase, hold prev max bin, dataMin = 1 dataMax = 0, ie move slow
+                binMax = lrintf(state->centerFreq[state->updateAxis] / fftResolution);
+            } else { // there was a max, look backwards for min, don't compare to bin 0
+                for (int i = binMax; i > 1; i--) {
```
i > 1 will skip not only bin 0, but also bin 1.
Is this intended?

ctzsnooze replied:
As I intended it, if i > 1, the lowest possible value for i would be 2, and i-1 is 1.
Then we can compare bin 2 to bin 1, but not any lower.
That is intended to stop any comparison between bin 1 and bin 0, and I think it needs to be like this, since bin 0 does not carry any AC (frequency related) information.

rav-rav commented:
"Bin start and max values checked and kept within limits (not use bin 0 or >15)"

Why not use bin 0? It should be perfectly fine.

rav-rav commented:
//1 updateTicks is only needed for update rates above 12kHz, so it can be removed.

// 2..Is 2k gyro loop likely to work?
It was designed to work even at 1kHz.

// 3.  What about Bosch gyro loop rates?  
What is special about those, what are the rates?

// 4.  does gyroDataAnalyseInit need to be called every time gyroDataAnalyseStateInit is called ??
Yes, because all values need to be recalculated based on the provided loop time. Where else would you like to call it?

Balint134 commented on src/main/flight/gyroanalyse.c:
```diff
@@ -112,13 +152,15 @@ void gyroDataAnalyseInit(uint32_t targetLooptimeUs)
+    dynNotchMaxCtrHz = fftSamplingRateHz * 0.48; // Stay below Nyquist; frequency at which each axis updates
 
-    dynNotchMaxCtrHz = fftSamplingRateHz / 2; //Nyquist
+    smoothFactor = 2 * 3.14f * DYN_NOTCH_SMOOTH_HZ / (gyroLoopRateHz / 12); // minimum PT1 k value
```
Is `3.14` meant to be the value of Pi? If so, you can use `M_PIf` from `maths.h`

ctzsnooze replied:
Thanks, it is.  It doesn't require much accuracy.  What would be computationally most efficient?

Balint134 replied:
Computationally they'd be the same, however semantically would be more correct nonetheless.

dogjutsu commented on src/main/flight/gyroanalyse.c:
```diff
@@ -279,6 +322,7 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state,
                     if (!fftIncreased) {
                         binStart = i; // first up-step bin
                         fftIncreased = true;
+                        dataMin = state->fftData[i - 1];
```
Looks like dataMin is only being set the once here (the very first time a bin is detected that is greater than the previous, ie when !fftIncreased) during the peak detection loop.  Since dataMax:dataMin ratio is important in this new implementation for dynamic control of the PT1 smoothing (deciding what a 'big' peak is), should dataMin be updated as the rest of the bins are checked whenever a new lower min is detected?

As is, I would think that if the low frequency end of the FFT happens to be relatively noisy then it could skew the ratio dataMax:dataMin ratio.  For example, imagine these imaginary bin contents:

A: [7,8,6,1,2,1,1,1,1,1,4,9,3,1,1,1,1...]
vs
B: [1,2,1,7,8,6,1,1,1,1,4,9,3,1,1,1,1...] 

Both of these have the same 3 'spikes':  [1,2,1], [7.8.6], and [4,9,3] with the rest of the bins having the value '1'.  For scenario 'A' the the dataMax:dataMin ratio would be 9:8, and for scenario 'B' it would be 9:1, and ideally I believe you'd want that ratio to be the same in both of the above cases as all the spikes are the same, just re-ordered.

I think probably it should keep checking for lower dataMins throughout the loop, something like this at the same level as the 'if (!fftIncreased)'?:

if (state->fftData[i - 1] < dataMin) {
    dataMin = state->fftData[i - 1];
}

dogjutsu commented on src/main/flight/gyroanalyse.c:
```diff
@@ -311,25 +311,30 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state,
+                    if (state->fftData[i] < state->fftData[i - 1]) { // up step below this one
+                        dataMin = state->fftData[i]; // set min if smaller exists before max
+                        break;
+                    }
```
I'm not sure how often this scenario might actually occur in real life or in the bigger picture how important it actually is, but I think this logic could potentially greatly under appreciate the magnitude of noise spikes that have a 'shape' where there is a false summit followed by a little valley just before the actual peak.

Let say a the noise floor in the FFT is '1', and the FFT bucket contents have a section like this, where '9' is the value in the binMax bucket (values completely made up and aren't floats, but relative value is what is relevant to the example):
[... 1, 1, 1, 8, 7, 9, 1, 1, 1 ...]

It would look backwards starting from the binMax (the one with the value '9') -- it would stop at dataMin = 7 because of the '8', and the dataMax:dataMin ratio used to set the dynamicFactor would be a modest 9:7, which could make the notch slower to target this spike if it suddenly popped up as a nasty resonance at a certain throttle value or something.

That leaves open what a 'better' way of selecting dataMin would be without getting too computationally expensive... perhaps have it only 'break' in the if after a minimum number of steps down from binMax (and only setting dataMin if the new value would be lower than the previous).

ctzsnooze replied:
It is very difficult.  Originally the code just got the first minimum it came across.  This method works better. 

In the example you provide, there are two peaks.  It would go to 9 because that was biggest. The weighting would set the actual filter location half way between 7 and 9, almost in the middle of the whole 8,7,9 peak.  It would stay on 9 indefinitely, until 9 became less than some other peak, even though it would perceive 9 as being a 'weak' peak, because it would always go to the biggest.  Should peak 9 fade, and become smaller than 8, the focus would shift to the 8 peak, and the move would be fairly quick.  But the centre would be also half-way through 7, or close to the middle of the whole lump, and wouldn't shift all that much.

Should the array be something like [... 1, 3, 7, 2, 1, 5, 9, 2, 1 ...] then we have two peaks, one big and one small.  It will go for 9, and stay there until either 9 drops below 7, when it would head to the peak at 7, fairly quickly.  If these peaks are similar in shape, it may jitter from one to the other.  Widely separated and similar height peaks like will always be a problem, though the code moves relatively quickly if they are large, and will favour staying on the most persistent.  If the moves from one to the other are delayed by smoothing, it will never stay on either of them.  That's why it has to move quickly to big peaks.

If the peak at 9 was resonance, and the pattern was [... 1, 3, 2, 1, 1, 5, 9, 2, 1 ...] the notch would go to 9, and would typically reduce the resonance, perhaps making the peak at 3 the largest after about 15-20ms.  But it would move only relatively slowly down towards 3 since that is a weak peak.  Resonance being resonance, as the notch moved away from the resonant point, the resonant peak would tend to return.  As it came back, the relatively slow drift of the FFT away from the resonant point would stop, and the notch would move back towards the main resonant peak.  It that resonant peak was fairly strong, it would not move far away before the resonance tended to come back and re-centre the filter.

I tested quite a few alternatives and this method of detecting max and min gave the best results.  If anyone can figure a better way, so we can do A-B testing, I'd welcome that.

ctzsnooze replied:
@dogjutsu - I will test looking forward from a peak to get a 'higher-bin min', as well as looking backwards, as I do now.  We can then use the lowest height of either of those.  That might better address the issue you raise than just looking backwards.

dogjutsu replied:
@ctzsnooze this sounds like a good solution to me -- should be more reliably representative of the relative amplitude of the peak.

ctzsnooze commented:
@rav-rav Thanks for your comments and clarifications.  I can remove some of my comments now.

The Bosch gyro sends data and is clocked 3200Hz I think.  Do you think that would work?

As for where to call gyroDataAnalyseInit, I had no clue about that, so I can't propose an alternative.  It did cross my mind that the others are called from gyroanalyse.h so I had wondered why it wasn't just listed there also.  But I don't have a technical understanding of these things, and am very happy to leave it where it is.  That comment will be removed now.  Thanks.

I set a debug to log bin0 and it had peculiar data in it compared to bin 1.  Since we have AC signals and bin0 is the DC component, bin0 is typically very small, and sometimes goes negative.  Bin 1 is our first AC data element and hence is the first bin that we can compare a higher bin to.  That's why I carefully made sure we never compare to bin0.  In practice bin 1 is nearly always bigger than bin2, so it is very uncommon to ever go looking down there.

ctzsnooze commented:
@rav-rav should this comment be kept?
line 205 `    // this runs at 1kHz`
I get the impression that gyroDataAnalyse runs at gyro loop rate?

rav-rav commented:
Ah, I was to quick to respond, there are some constrains!

Originally fftSamplingRateHz was always 1kHz
Line 205 and 208 are reflecting this and should be changed to fftSamplingRateHz instead of 1kHz.
gyroDataAnalyse does run at gyro rate but all gyro data is downsampled to the frequency set in fftSamplingRateHz.

In any case fftSamplingRateHz must be achievable by an integer division from samplingFrequency = 1000000 / targetLooptimeUs.

So for 4kHz all is fine as it is a whole number multiple of the selectable fft sampling frequencies of 1000, 1333 and 2000 for low/med/high.
For 2kHz only LOW and HIGH can be selected, otherwise the accumulation of samples won't work correctly and the results would be wrong.
Obviously for 1kHz sampling only DYN_NOTCH_RANGE_HZ_LOW can be used.

For Bosch sensors running at 3200Hz, these settings would also need to be adjusted.
DYN_NOTCH_RANGE_HZ_LOW could be 800, MED 1066 and HIGH 1600.

etracer65 commented:
Any ranges need to be set dynamically and not hardcoded for various loop times. For example, I'll soon be working on  driver for the BMI270 which runs at 6.4KHz. We also support the ICM20649 (although I don't know of any boards using it) which runs at 9KHz sampling.

rav-rav commented:
fftSamplingRateHz can be calculated by dividing the gyro sampling rate by a whole number.
Half of this rate is the maximum frequency the dynamic notch can reach. I think showing this range and allowing it to be set would be quite intuitive.

How about calculating possible denominators up to a certain frequency of the fft (for example 2,25kHz) and allowing these to be set?

For 4k gyro loop you could select 2, 3, 4. This would result in fft sampling of 2000, 1333, 1000 and an effective dynamic notch at frequencies up to 1000, 666, 500.
In the configurator there could be a slider showing those three ranges.

For 6.4k loop you could select 3, 4, 5, 6. This would result in fft sampling of 2133, 1600, 1280, 1066.
Here you would get a slider between 533 and 1066.

For 9k loop you could select 4, 5, 6, 7, 8, 9. This would result in fft sampling of 2250, 1800, 1500, 1286, 1125, 1000.
Here you would get a slider between 500 and 1225.

rav-rav commented:
@ctzsnooze did you ever try different window sizes for FFT_WINDOW_SIZE?
This was not possible with F3 processors. But F4 or F7 might be able to handle 64. It would be interesting to see if results get better with twice the resolution per bin.

ctzsnooze commented:
Implementing the range options for a given looptime, while making it simple for the user to select a lower vs a higher range... that sounds tricky.
Let's say the user chose an allowed divider value for a given looptime, but then chose a different gyro looptime, one for which the old divider value was inappropriate... that would cause issues.
Most users currently just leave it on medium.
Maybe we could retain low, medium and high, and just select, in the code, the nearest divider that reflects the traditional low, medium or high range.  For 1k only high would be possible; for 2k we could map low and medium to 1k sampling and high to 2k, and so on.  
I think I should leave this to people more clever than me lol.

ctzsnooze commented:
@rav-rav no I didn't personally try 64 bins, but @kmitchel did some tests, and yes, it should be quite possible.

dogjutsu commented on src/main/flight/gyroanalyse.c:
```diff
@@ -287,44 +331,39 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state,
-            state->centerFreq[state->updateAxis] = centerFreq;
+
+            // PT1 style dynamic smoothing moves rapidly towards big peaks and slowly away, up to 5x faster
+            float dynamicFactor = constrainf(dataMax / dataMin, 1.0f, 5.0f);
```
@ctzsnooze have you considered making the constraints on dynamicFactor (or perhaps just the high end constraint) tunable, either directly or with a meta-value(s) that impact the range?

I acknowledge that there are many good reasons to keep down the number of tunables, but even beyond different quads  / flight styles, things like the other filter settings (rpm vs not!) or gyro sample rate could make a big difference on the 'personality' of the gyro signal even on the same quad.  I suspect there *could* be benefit to be had from tuning the dynamicFactor range for different quads/filtering scenarios/etc.

Perhaps make it tunable for testing purposes initially (or a special build that does so) to make it accessible for testers to try different values and help determine whether it makes sense as a tunable or if there is a specific range that works best across a wide array of quads / setups better than 1.0 - 5.0.

ctzsnooze commented:
@rav-rav - could you please advise on how to bypass the Hanning window functionality, but still rotate the gyro data into the fftData?  

I tried commenting out all the hanning step lines but of course that failed.  

The reason I'm asking is not because I hope to reduce CPU, but because I wonder if it will provide more frequency stability, accepting that it may be worse in terms of amplitude assessment accuracy.  I have seen in logs with dual peaks that the FFT maxBin can jump from one peak to the other and back within a few FFT runs.  I had not expected that such rapid changes could be possible given that only one or two samples were exchanged in a 32 bit deep buffer.  

I wondered if the Hanning window, with so few samples in the buffer, may be focusing the frequency calculation on just a small number of data points in the centre of the buffer, rather than all data points equally.  I have a good model to test this hypothesis on.  But I can't figure how to disable Hanning windowing without breaking it - I'd be grateful for your help with that.

Balint134 commented:
@ctzsnooze set the hanning window to `1.0f` in `gyroDataAnalyseInit()`, then it will leave the gyro samples untouched. This way you can quickly test whatever you want :)

ctzsnooze commented:
Ah, very clever, so just:
```
    for (int i = 0; i < FFT_WINDOW_SIZE; i++) {
        hanningWindow[i] = 1.0f;
    }
```
Will try.

Asizon commented:
It works @Balint134! 
![image-3](https://user-images.githubusercontent.com/43983086/71958945-da71ec80-31f1-11ea-823e-d597180360a7.png)
Lets flight to explore it!

rav-rav commented:
@ctzsnooze did you look at the [wiki page](https://en.wikipedia.org/wiki/Window_function#Hann_.28Hanning.29_window) regarding window functions?
Rectangular just seems wrong to me, but there might be better functions than hanning for our use case.

It might be worth it to record raw gyro data and run this through different functions.

In your example bin 3 and 9 are probably very similar in amplitude regarding the very low timeframe and coarse resolution the fft is providing. Maybe it would help to use a hysteresis and only allow moving to a different frequency, if it is significantly higher than the current.

ctzsnooze commented:
@rav-rav I did research Hanning vs Rectangular windows quite a lot and testing so far shows no big differences.  
What I’m finding quite confusing at the moment is how quickly the ‘tallest’ FFT bin can change so quickly when I log it. 
In 8k medium mode we add one new value to the 32 sample FFT data buffer after every 6 gyro loops.  That value is the average of the preceding 6 gyro values for a given axis. Hence only one of the 32 values changes every ¾ of a millisecond. It will take 24ms to completelu

rav-rav commented:
@ctzsnooze you are correct regarding the circular buffer pointer and the start/end of current data. And I do realize it is unnecessary complicated, but I don't see any reason for it to have wrong results.

A less complicated way would be to copy the data in the correct order into fftData first and then apply the hanning window in a second step.

Here is how it works right now:

Let's assume we have a buffer for 10 datapoints and the center index is currently at idx=6 after sampling 16 datapoints starting at 0.
The circular buffer is looking like this:
circ = [10, 11, 12, 13, 14, 15, 6, 7, 8, 9]

idx is pointing at the oldest sampling point "6".

So the correct order for all data would be index 6-9 first followed by data from index 0-5.
This data has to be multiplied by the window function before going into fftData[].

This happens in two steps:

So we have fftData[0-3] = circ[6-9] * hanning[0-3] for the first 4 datapoints.
And fftData[4-9] = circ[0-5] * hanning[4-9] for the last 6 datapoints.

The index for the hanning window in the first step is always 0.
In the second step it is calculated from the windowsize. In this example it is 10 - 6 = 4.

ctzsnooze commented:
You know, this is where the truly amazing bits of coding are just pure genius stuff. Now you explain how those two lines work, it all makes perfect sense.  Thanks so much for how you put this together so elegantly.

rav-rav commented:
Well thanks, but I think good code should be easy to read and understand. So something could be improved here.

etracer65 commented on src/main/sensors/gyro.h:
```diff
@@ -144,13 +133,15 @@ typedef struct gyroConfig_s {
     uint16_t dyn_lpf_gyro_max_hz;
-    uint8_t  dyn_notch_range;            // ignore any FFT bin below this threshold
+
+    uint16_t dyn_notch_max_hz;
```
Since you're changing the structure of this (by changing from uint8 to uint16) you must increment the PG version otherwise the parameter group will be corrupted from saved settings. In gyro.c at about line 160 change:
```
PG_REGISTER_WITH_RESET_FN(gyroConfig_t, gyroConfig, PG_GYRO_CONFIG, 7);
```
to
```
PG_REGISTER_WITH_RESET_FN(gyroConfig_t, gyroConfig, PG_GYRO_CONFIG, 8);
```

etracer65 commented on src/main/msp/msp.c:
```diff
@@ -1651,7 +1651,7 @@ static bool mspProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst)
         // Added in MSP API 1.42
 #if defined(USE_GYRO_DATA_ANALYSE)
-        sbufWriteU8(dst, gyroConfig()->dyn_notch_range);
+        sbufWriteU8(dst, gyroConfig()->dyn_notch_max_hz);
```
`gyroConfig()->dyn_notch_max_hz` is a uint16 - not a uint8 any more. This can't be done this way. The MSP clients won't understand this and it will be a corrupted MSP message. Typically we leave the old value in place (probably just send a 0) and add the new value on to the end. But in this case need to think about what will happen for example if somebody uses a Configurator version that's not aware of this change and how it will behave. It won't know to read the extra new uint16 and that's probably ok. But what would it do with the hardcoded value for the old range?

etracer65 commented on src/main/msp/msp.c:
```diff
@@ -2462,12 +2462,12 @@ static mspResult_e mspProcessInCommand(mspDescriptor_t srcDesc, uint8_t cmdMSP,
             gyroConfigMutable()->dyn_notch_min_hz = sbufReadU16(src);
 #else
-            sbufReadU8(src);
+            sbufReadU16(src);
```
See comments above - can't change the structure of a MSP message this way.

ctzsnooze commented:
I've fixed the increment the PG version error, thanks.
For the MSP issues I'll have to be guided as to what's best to do.

mikeller commented on src/main/cli/settings.c:
```diff
@@ -658,10 +658,10 @@ const clivalue_t valueTable[] = {
     { "gyro_to_use",                VAR_UINT8  | HARDWARE_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_GYRO }, PG_GYRO_CONFIG, offsetof(gyroConfig_t, gyro_to_use) },
 #endif
 #if defined(USE_GYRO_DATA_ANALYSE)
-    { "dyn_notch_range",           VAR_UINT8   | MASTER_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_DYNAMIC_FILTER_RANGE }, PG_GYRO_CONFIG, offsetof(gyroConfig_t, dyn_notch_range) },
```
`TABLE_DYNAMIC_FILTER_RANGE` and `lookupTableDynamicFilterRange` have to be removed from `settings.h` and `settings.c` respectively.

mikeller commented on src/main/cms/cms_menu_imu.c:
```diff
@@ -630,7 +624,7 @@ static const OSD_Entry cmsx_menuDynFiltEntries[] =
 
 #ifdef USE_GYRO_DATA_ANALYSE
-    { "NOTCH RANGE",    OME_TAB,    NULL, &(OSD_TAB_t)    { &dynFiltNotchRange,   3, osdTableDynNotchRangeType}, 0 },
+    { "NOTCH MAX HZ",   OME_UINT16, NULL, &(OSD_UINT16_t) { &dynFiltNotchMaxHz,   0, 1000, 1 }, 0 },
```
Moving this down to below `"NOTCH MIN HZ"` will make the UI a bit cleaner.

mikeller commented on src/main/flight/gyroanalyse.c:
```diff
@@ -44,31 +44,59 @@
+// for LOW we get 16 x 31.25ms bins to 500hz, 
+// for HIGH we get 16 x 62.5ms bins to 1000Hz
+
+//#define FFT_WINDOW_SIZE 32  -> this is in gyroanalyse.h
```
Commented out lines of code should be removed for maintainability.

mikeller commented on src/main/flight/gyroanalyse.c:
```diff
@@ -82,46 +110,34 @@ void gyroDataAnalyseInit(uint32_t targetLooptimeUs)
 
     for (int i = 0; i < FFT_WINDOW_SIZE; i++) {
         hanningWindow[i] = (0.5f - 0.5f * cos_approx(2 * M_PIf * i / (FFT_WINDOW_SIZE - 1)));
+        // for testing, to disable, set to hanningWindow[i] = 1.0f;
```
If this is only for testing, should it be removed now?

mikeller commented on src/main/flight/gyroanalyse.c:
```diff
@@ -131,21 +147,20 @@ void gyroDataAnalyseStateInit(gyroAnalyseState_t *state, uint32_t targetLooptime
-    const uint16_t samplingFrequency = 1000000 / targetLooptimeUs;
-    state->maxSampleCount = samplingFrequency / fftSamplingRateHz;
-    state->maxSampleCountRcp = 1.f / state->maxSampleCount;
+//    const uint16_t samplingFrequency = 1000000 / targetLooptimeUs;
```
See above about commented out code.

mikeller commented on src/main/flight/gyroanalyse.c:
```diff
@@ -269,62 +286,76 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state,
+            float fftWeightedSum = squaredData * binMax;
+            // accumulate upper shoulder unless it would be FFT_BIN_COUNT
+            binMax += 1;
+            if (binMax < FFT_BIN_COUNT){
```
Missing space. :space_invader:

mikeller commented on src/main/flight/gyroanalyse.c:
```diff
@@ -269,62 +286,76 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state,
+            }
+            binMax -= 1;
+            // accumulate lower shoulder unless lower shoulder would be bin 0 (DC)
+            if (binMax > 1){
```
:space_invader:

mikeller replied:
Whitespace still missing.

mikeller commented on src/main/flight/gyroanalyse.h:
```diff
@@ -24,9 +24,7 @@
-
-// max for F3 targets
-#define FFT_WINDOW_SIZE 32
+#define FFT_WINDOW_SIZE 32 // Max for F3
```
F3 isn't that relevant any more, as it is no longer supported in Betaflight - should we consider increasing the window size?

mikeller commented on src/main/msp/msp.c:
```diff
@@ -1664,12 +1664,12 @@ static bool mspProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst)
         // Added in MSP API 1.42
 #if defined(USE_GYRO_DATA_ANALYSE)
-        sbufWriteU8(dst, gyroConfig()->dyn_notch_range);
+        sbufWriteU16(dst, gyroConfig()->dyn_notch_max_hz);
```
The correct way to do this in a backwards compatible way is:
- replace all lines in here dealing with `dyn_notch_range` with writes of 0 and empty reads respectively, and add comments like in https://github.com/betaflight/betaflight/pull/9327/files#diff-eefd576c86ccfa8b6cdf98081806a7a8L1690;
- add writes / reads of `dyn_notch_max_hz` at the end of the respective commands, in a section for API version 1.43, like what is done for 1.42 in https://github.com/betaflight/betaflight/pull/9327/files#diff-eefd576c86ccfa8b6cdf98081806a7a8L2475.

ctzsnooze commented:
Thanks, @mikeller  and @rav-rav ...
- removed `TABLE_DYNAMIC_FILTER_RANGE` and `lookupTableDynamicFilterRange`, apologies for not doing that clean up properly
-  removed comment `// Max for F3`.  Not sure where that came from.  Looks like I added it, but it wasn't something I recall doing.  The FFT sample width can indeed be 64, and will improve frequency resolution, but I have no experience with 64 samples.  Something to evaluate in the future perhaps.  
- moved `NOTCH MAX HZ` to below min in CMS
- revised, removed and tidied comments
- made a  variable `shoulderBin` instead of fiddling with `binMax` as per Rav's comment
- minor changes to smoothing defaults based on logs I've received.
- marked all comments as resolved, except msp.c, which I can't figure out what to do.

ctzsnooze commented:
@mikeller Thanks for your reviews.  I am happy that the content is good to go.
I have had a try at making the requested fixes to `msp.c`in the last commit, but something isn't right, it now freezes my FC if I change and save anything in the Filters tab of the Configurator.

ctzsnooze commented:
Increasing the buffer size by 1 in `msp.c` seemed sensible...
It stops the freezing... and Configurator doesn't show wonky numbers any more for Dynamic Notch... but now Configurator shows a bad value for rpm filters and won't save changes to either rpm or dynamic notch.
Well... .at least, I tried.
Let me know exactly what to do to fix it, and I will.

mikeller commented on src/main/flight/gyroanalyse.c:
```diff
@@ -333,26 +356,25 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state,
-                DEBUG_SET(DEBUG_FFT_FREQ, 1, state->centerFreq[state->updateAxis]);
-            }
-            // Debug FFT_Freq carries raw gyro, gyro after first filter set, FFT centre for roll and for pitch
+//            if (state->updateAxis == 1) {
```
Unused commented out code should be removed.
